File: Two Pointers/125. Valid Palindrome/Solution.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Time Complexity : 0(N)
        # Space Complexity : O(N)
        # res = ''
        # for letter in s:
        #     if letter.isalnum():
        #         res += letter.lower()
        
        # return res == res[::-1]
        
        # Time Complexity : 0(N)
        # Space Complexity : O(1)
        l, r = 0, len(s) - 1

        while l < r:
            left = s[l].lower()
            right = s[r].lower()
            if not left.isalnum():
                l += 1
                continue
            if not right.isalnum():
                r -= 1
                continue
            
            if left != right:
                return False
            l += 1
            r -= 1
        return True

File: Two Pointers/125. Valid Palindrome/Solution.java
class Solution {
    public boolean isPalindrome(String s) {
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        // StringBuffer str = new StringBuffer();
    
        // for (int i=0; i<s.length(); i++){
        //     char c = Character.toLowerCase(s.charAt(i));
        //     if (Character.isLetterOrDigit(c)){
        //         str.append(c);
        //     }
        // }
        // s = str.toString();
        // String reversedStr = str.reverse().toString();
        // return s.equals(reversedStr);

        
        // Time Complexity: O(N)
        // Space Complexity: O(1)
        int l = 0;
        int r = s.length() - 1;
        while (l < r){
            char left = Character.toLowerCase(s.charAt(l));
            char right = Character.toLowerCase(s.charAt(r));
            if (Character.isLetterOrDigit(left) == false) {
                l++;
                continue;
            }

            if (Character.isLetterOrDigit(right) == false){
                r--;
                continue;
            }
            
            if (left != right) return false;
            l++;
            r--;
        }
        return true;
    }
}